crates/bevy_transform/src/hierarchy/child_builder.rs

```diff
@@ -11,25 +37,21 @@ pub struct InsertChildren {
 
 impl Command for InsertChildren {
     fn write(self: Box<Self>, world: &mut World, _resources: &mut Resources) {
-        for child in self.children.iter() {
-            world
-                .insert(*child, (Parent(self.parent), PreviousParent(self.parent)))
-                .unwrap();
-        }
-        {
-            let mut added = false;
-            if let Ok(mut children) = world.get_mut::<Children>(self.parent) {
-                children.0.insert_from_slice(self.index, &self.children);
-                added = true;
-            }
+        update_parent_and_previous_parent(self.parent, &self.children, world);
+
+        if let Ok(mut new_children) = world.get_mut::<Children>(self.parent) {
+            let index = self.index.min(new_children.len());
 
-            // NOTE: ideally this is just an else statement, but currently that _incorrectly_ fails borrow-checking
-            if !added {
-                world
-                    .insert_one(self.parent, Children(self.children))
-                    .unwrap();
+            for child in self.children.iter().rev() {
+                if !new_children.0.contains(child) {
```

Comment:
Here and on line 75 you're scanning through the existing children once for every new child. I think that's worse than the previous version where you built a HashSet. That required a scan through the existing children once per command application. Though this version might actually be faster in the case where there's only a single child being added but I'd expect it to be much slower if there are multiple new children.